File: Src/java/cql-to-elm/src/jvmTest/java/org/cqframework/cql/cql2elm/operators/CqlListOperatorsTest.java
package org.cqframework.cql.cql2elm.operators;

import static kotlinx.io.CoreKt.buffered;
import static kotlinx.io.JvmCoreKt.asSource;
import static org.cqframework.cql.cql2elm.matchers.HasTypeAndResult.hasTypeAndResult;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import org.cqframework.cql.cql2elm.CqlTranslator;
import org.cqframework.cql.cql2elm.LibraryManager;
import org.cqframework.cql.cql2elm.ModelManager;
import org.hl7.elm.r1.*;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

/**
 * Created by Bryn on 12/30/2016.
 */
class CqlListOperatorsTest {
    // NOTE: The CQL for this test is taken from an engine testing suite that produced a particular issue with ambiguous
    // conversions.

    private static Map<String, ExpressionDef> defs;

    @BeforeAll
    static void setup() throws IOException {
        ModelManager modelManager = new ModelManager();
        CqlTranslator translator = CqlTranslator.fromSource(
                buffered(asSource(CqlListOperatorsTest.class.getResourceAsStream("../OperatorTests/CqlListOperators.cql"))),
                new LibraryManager(modelManager));
        assertThat(translator.getErrors().size(), is(0));
        Library library = translator.toELM();
        defs = new HashMap<>();
        if (library.getStatements() != null) {
            for (ExpressionDef def : library.getStatements().getDef()) {
                defs.put(def.getName(), def);
            }
        }
    }

    @Test
    void union() {
        ExpressionDef def = defs.get("Union123AndEmpty");
        assertThat(def, hasTypeAndResult(Union.class, "list<System.Integer>"));
    }
}

File: Src/java/cql-to-elm/src/jvmTest/kotlin/org/cqframework/cql/cql2elm/operators/CqlListOperatorsTest.kt
package org.cqframework.cql.cql2elm.operators

import java.io.IOException
import kotlinx.io.asSource
import kotlinx.io.buffered
import org.cqframework.cql.cql2elm.CqlTranslator.Companion.fromSource
import org.cqframework.cql.cql2elm.LibraryManager
import org.cqframework.cql.cql2elm.ModelManager
import org.cqframework.cql.cql2elm.matchers.HasTypeAndResult.Companion.hasTypeAndResult
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.`is`
import org.hl7.elm.r1.ExpressionDef
import org.hl7.elm.r1.Union
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

/** Created by Bryn on 12/30/2016. */
internal class CqlListOperatorsTest {
    @Test
    fun union() {
        val def: ExpressionDef? = defs!!["Union123AndEmpty"]
        assertThat<ExpressionDef?>(def, hasTypeAndResult(Union::class.java, "list<System.Integer>"))
    }

    companion object {
        // NOTE: The CQL for this test is taken from an engine testing suite that produced a
        // particular issue with ambiguous
        // conversions.
        private var defs: MutableMap<String?, ExpressionDef?>? = null

        @JvmStatic
        @BeforeAll
        @Throws(IOException::class)
        fun setup() {
            val modelManager = ModelManager()
            val translator =
                fromSource(
                    CqlListOperatorsTest::class
                        .java
                        .getResourceAsStream("../OperatorTests/CqlListOperators.cql")!!
                        .asSource()
                        .buffered(),
                    LibraryManager(modelManager),
                )
            assertThat(translator.errors.size, `is`(0))
            val library = translator.toELM()
            defs = HashMap()
            if (library!!.statements != null) {
                for (def in library.statements!!.def) {
                    defs!![def.name] = def
                }
            }
        }
    }
}
